app/src/main/java/tech/relaycorp/courier/domain/StoreMessage.kt

```diff
@@ -11,29 +11,34 @@ import tech.relaycorp.courier.data.model.StoredMessage
 import tech.relaycorp.courier.data.network.Cargo
 import tech.relaycorp.courier.data.network.CargoCollectionAuthorization
 import tech.relaycorp.courier.data.network.RAMFMessage
-import java.io.InputStream
 import java.util.Date
 import javax.inject.Inject
 
 class StoreMessage
 @Inject constructor(
     private val storedMessageDao: StoredMessageDao,
-    private val diskRepository: DiskRepository
+    private val diskRepository: DiskRepository,
+    private val getStorageUsage: GetStorageUsage
 ) {
 
-    suspend fun storeCargo(data: InputStream) =
-        storeMessage(Cargo.wrap(data), MessageType.Cargo)
+    suspend fun storeCargo(cargo: Cargo) =
+        storeMessage(cargo, MessageType.Cargo)
 
-    suspend fun storeCCA(data: InputStream) =
-        storeMessage(CargoCollectionAuthorization.wrap(data), MessageType.Cargo)
+    suspend fun storeCCA(cca: CargoCollectionAuthorization) =
+        storeMessage(cca, MessageType.CCA)
+
+    private suspend fun storeMessage(message: RAMFMessage, type: MessageType): StoredMessage? {
+        if (!checkForAvailableSpace(message)) return null
 
-    private suspend fun storeMessage(message: RAMFMessage, type: MessageType): StoredMessage {
         val storagePath = diskRepository.writeMessage(message.payload)
```

Comment:
I missed this before, but we need to store the RAMF message instead of its payload. Also keep in mind you can't get the original bytes for the message from a `RAMFMessage` instance.

In other words, `storeMessage()` has to take an `InputStream` instead of a `RAMFMessage`.

Reply:
I was actually assuming `payload` would have everything inside, basically `RAMFMessage` would keep a copy of itself serialized. I wanted to avoid having a different serialization/deserialization going on.

Reply:
Probably the `payload` name is the issue here, because there's already a `payload` in the `RAMFMessage`.  Probably I mean a serializedBytes or something.

Reply:
Simply put, RAMF messages are containers with some metadata attached to them. _Payload_ refers to the thing the message contains. For example, the payload of a cargo message is a series of parcels and other messages in encrypted form; the payload of a parcel is the service message (for example, a JSON-serialised tweet) in encrypted form.

RAMF messages don't contain their own serialization (`ByteArray`/`InputStream`) because it isn't always available (when you're creating the message) and, most importantly, because it'd be inefficient memory-wise: If we kept a pointer to the original ByteArray, we'd prevent the garbage collector from reclaiming that space even when the app may no longer have any need for the original serialisation.

The courier app is a bit different from the other apps using the JVM library in that, even though it has to parse RAMF messages to validate them, it's ultimately interested in storing the original byte sequence. But gateways and endpoints will want to discard the original byte sequence once they've parsed the message because they're ultimately interested in the payload only.